src/lib/authselect_activate.c

```diff
@@ -30,6 +30,70 @@
 #include "lib/authselect_paths.h"
 #include "lib/authselect_util.h"
 
+static char *get_nsswitch_dir()
+{
+    static const char *filename = AUTHSELECT_NSSWITCH_CONF;
+    const char *slash;
+    char *path;
+
+    slash = strrchr(filename, '/');
+    if (slash == NULL) {
+        ERROR("Invalid location to nsswitch.conf: %s", filename);
+        return NULL;
+    }
+
+    if (slash == filename) {
+        /* It is placed in root directory.
+         * Weird, but technically acceptable. */
+        return "/";
+    }
+
+    path = strndup(filename, slash - filename);
+    if (path == NULL) {
+        ERROR("Out of memory!");
+    }
+
+    return path;
+}
+
+static bool
+check_directories()
+{
+    const char *dirs[] = {
+        AUTHSELECT_CONFIG_DIR,
+        AUTHSELECT_PAM_DIR,
+        AUTHSELECT_DCONF_DIR,
+        AUTHSELECT_DCONF_DIR "/locks",
+        NULL, /* place for nsswitch direcotry */
+        NULL
+    };
+    errno_t ret;
+    bool bret;
+    int i;
+
+    dirs[4] = get_nsswitch_dir();
+    if (dirs[4] == NULL) {
+        ERROR("Unable to get path to nsswitch.conf parent directory!");
+        return false;
+    }
+
+    bret = true;
+    for (i = 0; dirs[i] != NULL; i++) {
+        ret = check_access(dirs[i], W_OK | X_OK);
```

Comment:
So, why do we need this? In general, I'm not a big fan of testing something before action. I think it is preferable to test the action and fail gracefully. Because the DAC access control (permissions) doesn't have to be the only thing that prevents the code from running, later on MAC (=SELinux, AppArrmor, seccomp in containers, ..) can kick you out as well.

I mean, as long as it's about a nice error message, then fine, but I still think handling EPERM gracefully is a better design.